File: internal/clickhouse/clickhouse.go
package clickhouse

import (
	"context"
	"database/sql"
	"time"

	"github.com/ClickHouse/clickhouse-go/v2"
	"github.com/freshworks/load-generator/internal/stats"
	"github.com/sirupsen/logrus"
)

type Generator struct {
	DB *sql.DB

	log   *logrus.Entry
	o     GeneratorOptions
	ctx   context.Context
	stats *stats.Stats
}

type GeneratorOptions struct {
	DSN   string
	Query string
}

func NewOptions() *GeneratorOptions {
	return &GeneratorOptions{
		DSN:   "clickhouse://127.0.0.1:9000/default",
		Query: "SELECT 1",
	}
}

func NewGenerator(id int, options GeneratorOptions, ctx context.Context, requestrate int, s *stats.Stats) *Generator {
	log := logrus.WithFields(logrus.Fields{"Id": id})
	return &Generator{log: log, o: options, ctx: ctx, stats: s}
}

func (g *Generator) Init() error {
	var err error
	g.DB, err = g.Open(g.o.DSN)
	if err != nil {
		return err
	}

	return g.DB.Ping()
}

func (g *Generator) InitDone() error {
	return nil
}

func (g *Generator) Tick() error {
	start := time.Now()
	res, err := g.DB.QueryContext(g.ctx, g.o.Query)
	if err != nil {
		g.log.Errorf("ClickHouse error: %v", err)
		return err // Return the error so it's properly handled
	}

	err = res.Close()
	if err != nil {
		g.log.Errorf("ClickHouse close error: %v", err)
		return err
	}

	// Record successful query metrics
	var traceInfo stats.TraceInfo
	traceInfo.Type = stats.ClickHouseTrace
	traceInfo.Key = g.o.DSN // Use DSN as the key
	traceInfo.Subkey = g.o.Query
	traceInfo.Total = time.Since(start)
	g.stats.RecordMetric(&traceInfo)

	return nil
}

func (g *Generator) Finish() error {
	if g.DB != nil {
		return g.DB.Close()
	}

	return nil
}

func (g *Generator) Open(dsn string) (*sql.DB, error) {
	// Parse DSN and create ClickHouse connection
	opt, err := clickhouse.ParseDSN(dsn)
	if err != nil {
		return nil, err
	}

	return clickhouse.OpenDB(opt), nil
}

File: scripts/clickhouse.lua
local clickhouse = require('clickhouse')

-- 100 SQL calls in one tick and compute metrics for batch of 100 calls as
-- single unit

local db = nil

function init()
   local o = clickhouse.Options()
   o.DSN = "clickhouse://127.0.0.1:9000/default"

   local clickhouse_client, err = clickhouse.New(o)
   if err ~= nil then
      Log:Errorf("%v", err)
      return err
   end

   db = clickhouse_client.DB
end

function tick()
   return exec_sql_queries(queries)
end

function exec_sql_queries(queries)
   local metric_name = "my_clickhouse_req_with_100_queries"
   LG:BeginCustomMetrics(metric_name)
   for num = 1,100 do
      local r, err = db:Query("SELECT 1")
      if err then
         Log:Errorf("Error %v %v\n", err, r)
         LG:EndCustomMetricsWithError(metric_name)
         return
      end

      local err = r:Close()
      if err then
         Log:Errorf("%v", err)
      end
   end

   LG:EndCustomMetrics(metric_name)

   return
end
